File: packages/ruby/lib/crumbtrail/sender.rb
require 'net/http'
require 'uri'
require 'json'
require_relative 'event'

module Crumbtrail
  class Sender
    # 429 and 5xx are the only answers worth repeating the identical batch into. A 404 names a
    # session the cloud does not have, and every later batch for that id is refused the same
    # way, so retrying it four times only delays the gap that says the evidence is gone.
    RETRYABLE = [429].freeze
    # Shed reasons authored by the capture edge. An unrecognised reason is recorded as a plain
    # delivery failure rather than passed through as an invented classification.
    SHED_REASONS = %w[kill_switch sessions_per_hour bytes_per_day rate_limited_ingest
                      rate_limited_session_start trial_expired payment_failed upgrade_required].freeze
    EVENTS_PATH = '/api/events'.freeze
    ATTEMPTS = 4
    MAX_SHED_SECONDS = 300
    MAX_RESPONSE_BYTES = 4096

    def initialize(endpoint:, key:, cert_store: nil, logger: nil)
      @endpoint = URI(endpoint)
      raise ArgumentError, 'Crumbtrail requires HTTPS without credentials, query or fragment' unless @endpoint.is_a?(URI::HTTPS) && @endpoint.host && !@endpoint.userinfo && !@endpoint.query && !@endpoint.fragment
      raise ArgumentError, 'Crumbtrail ingest key is required' if key.to_s.strip.empty? || key.match?(/[[:cntrl:]]/)
      @key = key
      @cert_store = cert_store
      @logger = logger
      @queue = SizedQueue.new(64)
      @monitor = Mutex.new
      @wake = ConditionVariable.new
      @stopping = false
      @shed_until = nil
      @shed = nil
      @worker = Thread.new do
        while (batch = @queue.pop)
          begin
            deliver(batch[0], batch[1], batch[2])
          rescue Exception
            # A worker killed by anything other than a StandardError would leave `close` joining
            # a dead thread and reporting a clean drain over an empty queue.
            nil
          end
        end
      end
      @worker.report_on_exception = false
    end

    def enqueue(batch)
      session = (batch[:sessionId] || batch['sessionId']).to_s
      events = batch[:events] || batch['events'] || []
      @queue.push([session, JSON.generate(batch), events.size], true)
      true
    rescue ThreadError, ClosedQueueError, JSON::GeneratorError
      false
    end

    # Closing the queue allows queued batches to drain within the caller's deadline. Past the
    # deadline the worker is cancelled, so a retry backoff cannot hold process exit open past
    # the timeout the caller asked for.
    def close(timeout: 5)
      @queue.close
      return true if @worker.join(timeout)
      cancel
      @worker.join(1)
      false
    end

    private

    def cancel
      @monitor.synchronize do
        @stopping = true
        @wake.broadcast
      end
    end

    def stopping?
      @monitor.synchronize { @stopping }
    end

    # Interruptible backoff. `sleep` here blocks process exit for the whole retry budget.
    def pause(seconds)
      @monitor.synchronize do
        next if @stopping
        @wake.wait(@monitor, seconds)
      end
    end

    def deliver(session, payload, count, gap: false)
      return if stopping?
      if shedding?
        @shed[:dropped] += count unless gap
        return
      end
      flush_shed_gap unless gap
      ATTEMPTS.times do |attempt|
        return if stopping?
        response = begin
          post(payload)
        rescue StandardError
          nil # No response at all. A network failure is retried per the queue policy.
        end
        if response
          code = response.code.to_i
          if code == 202 && (directive = shed_directive(response))
            begin_shed(session, directive, count) unless gap
            return
          end
          return if code / 100 == 2
          unless RETRYABLE.include?(code) || code >= 500
            refuse(session, count, code) unless gap
            return
          end
        end
        break if attempt == ATTEMPTS - 1
        pause(0.25 * (attempt + 1))
      end
      refuse(session, count, nil) unless gap
    end

    def post(payload)
      uri = @endpoint.dup
      uri.path = EVENTS_PATH
      request = Net::HTTP::Post.new(uri)
      request['Authorization'] = "Bearer #{@key}"
      request['Content-Type'] = 'application/json'
      request.body = payload
      options = { use_ssl: true, open_timeout: 2, read_timeout: 3, write_timeout: 3 }
      options[:cert_store] = @cert_store if @cert_store
      Net::HTTP.start(uri.host, uri.port, **options) { |http| http.request(request) }
    end

    # A 202 passes any "is this a success" test while the cloud has already discarded the
    # evidence. Reading the body is the only way to tell the two apart.
    def shed_directive(response)
      body = response.body.to_s
      return nil if body.empty? || body.bytesize > MAX_RESPONSE_BYTES
      parsed = JSON.parse(body)
      return nil unless parsed.is_a?(Hash) && parsed['capture'] == 'shed'
      seconds = parsed['retryAfterSeconds']
      seconds = response['Retry-After'].to_i unless seconds.is_a?(Numeric)
      { reason: SHED_REASONS.include?(parsed['reason']) ? parsed['reason'] : 'delivery_failed',
        seconds: seconds.to_f.clamp(0, MAX_SHED_SECONDS) }
    rescue JSON::ParserError, TypeError
      nil
    end

    def begin_shed(session, directive, count)
      @shed = { session: session, reason: directive[:reason], dropped: count }
      @monitor.synchronize { @shed_until = Process.clock_gettime(Process::CLOCK_MONOTONIC) + directive[:seconds] }
      log("Crumbtrail capture is shed at the endpoint (#{directive[:reason]}); pausing delivery for #{directive[:seconds].round}s")
    end

    def shedding?
      deadline = @monitor.synchronize { @shed_until }
      !deadline.nil? && Process.clock_gettime(Process::CLOCK_MONOTONIC) < deadline
    end

    # The gap for a shed window cannot be delivered during the window. It is held and sent on
    # the first batch after Retry-After has passed.
    def flush_shed_gap
      return if @shed.nil? || shedding?
      shed = @shed
      @shed = nil
      @monitor.synchronize { @shed_until = nil }
      send_gap(shed[:session], shed[:reason], shed[:dropped], 'capture shed by the endpoint')
    end

    def refuse(session, count, code)
      detail = code ? "HTTP #{code}" : 'retry budget exhausted'
      log("Crumbtrail refused #{count} captured event(s): #{detail}")
      send_gap(session, 'delivery_failed', count, detail)
    end

    def send_gap(session, reason, count, detail)
      return if session.to_s.empty? || count.to_i <= 0
      event = Crumbtrail.event(Crumbtrail.now, 'capture_gap',
                               { kind: 'capture_gap', surface: 'backend_request', reason: reason,
                                 sessionId: session, droppedEventCount: count, detail: detail })
      deliver(session, JSON.generate(sessionId: session, events: [event]), 1, gap: true)
    rescue StandardError
      nil
    end

    def log(message)
      @logger ? @logger.warn(message) : warn(message)
    rescue StandardError
      nil
    end
  end
end

File: packages/go/sender.go
package crumbtrail

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"io"
	"log"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"sync"
	"time"
	"unicode"
)

type SenderConfig struct {
	Endpoint, Key string
	HTTPClient    *http.Client
	// Logger receives one line per permanent refusal and per capture shed window. Nil uses the
	// standard logger. A refusal that is only recorded as a capture gap is invisible until
	// somebody opens the session, and a revoked key deserves to be visible at the process.
	Logger *log.Logger
}

type Sender struct {
	endpoint, key string
	client        *http.Client
	logger        *log.Logger
	queue         chan pendingBatch
	done          chan struct{}
	ctx           context.Context
	cancel        context.CancelFunc
	mu            sync.RWMutex
	closed        bool
	// Owned by the single delivery goroutine.
	shedUntil   time.Time
	shedReason  string
	shedSession string
	shedDropped int
}

type pendingBatch struct {
	session string
	payload []byte
	events  int
	gap     bool
}

const senderAttempts = 4
const maxShedSeconds = 300
const maxResponseBytes = 4096

// shedReasons are authored by the capture edge. An unrecognised reason is recorded as a plain
// delivery failure rather than passed through as an invented classification.
var shedReasons = map[string]bool{"kill_switch": true, "sessions_per_hour": true, "bytes_per_day": true,
	"rate_limited_ingest": true, "rate_limited_session_start": true, "trial_expired": true,
	"payment_failed": true, "upgrade_required": true}

func NewSender(config SenderConfig) (*Sender, error) {
	u, err := url.Parse(config.Endpoint)
	if err != nil || u.Scheme != "https" || u.Host == "" || u.User != nil || u.RawQuery != "" || u.Fragment != "" || strings.TrimSpace(config.Key) == "" || strings.IndexFunc(config.Key, unicode.IsControl) >= 0 {
		return nil, errors.New("crumbtrail requires an HTTPS endpoint without credentials, query or fragment and a nonempty ingest key")
	}
	u.Path = "/api/events"
	u.RawPath = ""
	client := http.Client{Timeout: 5 * time.Second}
	if config.HTTPClient != nil {
		client = *config.HTTPClient
		client.Timeout = 5 * time.Second
	}
	client.CheckRedirect = func(*http.Request, []*http.Request) error { return http.ErrUseLastResponse }
	logger := config.Logger
	if logger == nil {
		logger = log.Default()
	}
	ctx, cancel := context.WithCancel(context.Background())
	s := &Sender{endpoint: u.String(), key: config.Key, client: &client, logger: logger,
		queue: make(chan pendingBatch, 64), done: make(chan struct{}), ctx: ctx, cancel: cancel}
	go func() {
		defer close(s.done)
		for {
			select {
			case <-ctx.Done():
				return
			case batch, ok := <-s.queue:
				if !ok {
					return
				}
				s.deliver(batch)
			}
		}
	}()
	return s, nil
}

func (s *Sender) Enqueue(batch Batch) bool {
	payload, err := json.Marshal(batch)
	if err != nil {
		return false
	}
	s.mu.RLock()
	defer s.mu.RUnlock()
	if s.closed {
		return false
	}
	select {
	case s.queue <- pendingBatch{session: batch.SessionID, payload: payload, events: len(batch.Events)}:
		return true
	default:
		return false
	}
}

// Close drains pending batches until ctx expires, then cancels network work.
func (s *Sender) Close(ctx context.Context) error {
	s.mu.Lock()
	if !s.closed {
		s.closed = true
		close(s.queue)
	}
	s.mu.Unlock()
	select {
	case <-s.done:
		s.cancel()
		return nil
	case <-ctx.Done():
		s.cancel()
		return ctx.Err()
	}
}

func (s *Sender) deliver(batch pendingBatch) {
	if s.shedding() {
		if !batch.gap {
			s.shedDropped += batch.events
		}
		return
	}
	if !batch.gap {
		s.flushShedGap()
	}
	for attempt := 0; attempt < senderAttempts; attempt++ {
		req, err := http.NewRequestWithContext(s.ctx, http.MethodPost, s.endpoint, bytes.NewReader(batch.payload))
		if err != nil {
			return
		}
		req.Header.Set("Authorization", "Bearer "+s.key)
		req.Header.Set("Content-Type", "application/json")
		response, err := s.client.Do(req)
		if err == nil {
			body, _ := io.ReadAll(io.LimitReader(response.Body, maxResponseBytes))
			response.Body.Close()
			if response.StatusCode == http.StatusAccepted {
				if reason, seconds, shed := shedDirective(response, body); shed {
					if !batch.gap {
						s.beginShed(batch, reason, seconds)
					}
					return
				}
			}
			if response.StatusCode/100 == 2 {
				return
			}
			// 429 and 5xx are the only answers worth repeating the identical batch into. A 404
			// names a session the cloud does not have, and every later batch for that id is
			// refused the same way, so retrying it only delays the gap that says the evidence
			// is gone.
			if response.StatusCode != http.StatusTooManyRequests && response.StatusCode < 500 {
				s.refuse(batch, response.StatusCode)
				return
			}
		}
		if attempt < senderAttempts-1 {
			select {
			case <-s.ctx.Done():
				return
			case <-time.After(time.Duration(attempt+1) * 250 * time.Millisecond):
			}
		}
	}
	s.refuse(batch, 0)
}

// A 202 passes any "is this a success" test while the cloud has already discarded the evidence.
// Reading the body is the only way to tell the two apart.
func shedDirective(response *http.Response, body []byte) (string, time.Duration, bool) {
	var directive struct {
		Capture           string  `json:"capture"`
		Reason            string  `json:"reason"`
		RetryAfterSeconds float64 `json:"retryAfterSeconds"`
	}
	if err := json.Unmarshal(body, &directive); err != nil || directive.Capture != "shed" {
		return "", 0, false
	}
	seconds := directive.RetryAfterSeconds
	if seconds <= 0 {
		if header, err := strconv.Atoi(strings.TrimSpace(response.Header.Get("Retry-After"))); err == nil {
			seconds = float64(header)
		}
	}
	if seconds < 0 {
		seconds = 0
	}
	if seconds > maxShedSeconds {
		seconds = maxShedSeconds
	}
	reason := directive.Reason
	if !shedReasons[reason] {
		reason = "delivery_failed"
	}
	return reason, time.Duration(seconds * float64(time.Second)), true
}

func (s *Sender) beginShed(batch pendingBatch, reason string, wait time.Duration) {
	s.shedUntil = time.Now().Add(wait)
	s.shedReason = reason
	s.shedSession = batch.session
	s.shedDropped = batch.events
	s.logger.Printf("crumbtrail: capture is shed at the endpoint (%s); pausing delivery for %s", reason, wait)
}

func (s *Sender) shedding() bool {
	return !s.shedUntil.IsZero() && time.Now().Before(s.shedUntil)
}

// The gap for a shed window cannot be delivered during the window. It is held and sent on the
// first batch after Retry-After has passed.
func (s *Sender) flushShedGap() {
	if s.shedUntil.IsZero() {
		return
	}
	session, reason, dropped := s.shedSession, s.shedReason, s.shedDropped
	s.shedUntil, s.shedReason, s.shedSession, s.shedDropped = time.Time{}, "", "", 0
	s.sendGap(session, reason, dropped, "capture shed by the endpoint")
}

func (s *Sender) refuse(batch pendingBatch, status int) {
	if batch.gap {
		return
	}
	detail := "retry budget exhausted"
	if status != 0 {
		detail = "HTTP " + strconv.Itoa(status)
	}
	s.logger.Printf("crumbtrail: refused %d captured event(s): %s", batch.events, detail)
	s.sendGap(batch.session, "delivery_failed", batch.events, detail)
}

func (s *Sender) sendGap(session, reason string, dropped int, detail string) {
	if session == "" || dropped <= 0 {
		return
	}
	event := newEvent(time.Now().UnixMilli(), "capture_gap", map[string]any{
		"kind": "capture_gap", "surface": "backend_request", "reason": reason,
		"sessionId": session, "droppedEventCount": dropped, "detail": detail})
	payload, err := json.Marshal(Batch{session, []Event{event}})
	if err != nil {
		return
	}
	s.deliver(pendingBatch{session: session, payload: payload, events: 1, gap: true})
}
